### PR TITLE
New apply external field excitation on grid

### PR DESCRIPTION
**Pull request by jackieyao0114**
1. Distinguished the application of each field component by defining the flag `which_variable` in function `ApplyExternalFieldExcitationOnGrid`;
2. Modified the sequence of applying external field excitation in `WarpXEvolve.cpp`.

- New sequence: apply excitation on E, B and H (if `USE_LLG=TRUE`) -> EvolveB & FillBoundaryB -> apply excitation on B ->  EvolveHM & FillBoundaryHM & apply excitation on H (if `USE_LLG=TRUE`) -> EvolveE & FillBoundaryE -> apply excitation on E -> EvolveB & FillBoundaryB -> apply excitation on B ->  EvolveHM & FillBoundaryHM & apply excitation on H (if `USE_LLG=TRUE`)

### CODE REVIEW

File: Source/FieldSolver/WarpXExternalEMFields.cpp [hunk 1]
```diff
@@ -3,43 +3,53 @@
...
+ * \brief which_variable determines which field component the external excitation is applied on
+ * which_variable = 0 : external field excitation applied to all three field components, E, B and H
+ * which_variable = 1 : external field excitation applied to E field only
+ * which_variable = 2 : external field excitation applied to B field only
+ * which_variable = 3 : external field excitation applied to H field only; this option is only valid when USE_LLG == TRUE
+ */
+
 void
-WarpX::ApplyExternalFieldExcitationOnGrid ()
+WarpX::ApplyExternalFieldExcitationOnGrid (std::string const& fieldtype)
```
Comment (jackieyao0114) [inline Source/FieldSolver/WarpXExternalEMFields.cpp]: I’m not so confident in the input variable I have in `WarpX::ApplyExternalFieldExcitationOnGrid (std::string const& fieldtype)`  , does the constant string input make sense?

Comment (RevathiJambunathan) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: Its better to use
`int const fieldtype`

File: Source/FieldSolver/WarpXExternalEMFields.cpp [hunk 2]
```diff
@@ -3,43 +3,53 @@
...
-                                               Efield_fp[lev][1].get(),
-                                               Efield_fp[lev][2].get(),
-                                               getParser(Exfield_xt_grid_parser),
-                                               getParser(Eyfield_xt_grid_parser),
-                                               getParser(Ezfield_xt_grid_parser),
-                                               getParser(Exfield_flag_parser),
-                                               getParser(Eyfield_flag_parser),
-                                               getParser(Ezfield_flag_parser),
-                                               lev );
+        if (fieldtype == "all" || fieldtype == "efieldexternal") {
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXExternalEMFields.cpp]:
```suggestion
        if (fieldtype == ExternalFieldType::All || fieldtype == ExternalFieldType::EfieldExternal") {
```

Comment (RevathiJambunathan) [reply Source/FieldSolver/WarpXExternalEMFields.cpp]: and similarly for the other fields

File: Source/Evolve/WarpXEvolve.cpp [hunk 3]
```diff
@@ -345,8 +345,11 @@ WarpX::OneStep_nosub (Real cur_time)
     if (do_pml && pml_has_particles) CopyJPML();
     if (do_pml && do_pml_j_damping) DampJPML();
 
-    // ApplyExternalFieldExcitation
-    ApplyExternalFieldExcitationOnGrid();
+    if (cur_time == 0._rt) { // at the first time step, make sure to apply the hard source before fields get evolved
+        // ApplyExternalFieldExcitation
+        ApplyExternalFieldExcitationOnGrid("all");
```
Comment (RevathiJambunathan) [inline Source/Evolve/WarpXEvolve.cpp]:
```suggestion
        ApplyExternalFieldExcitationOnGrid(ExternalFieldType::All);
```

Comment (RevathiJambunathan) [reply Source/Evolve/WarpXEvolve.cpp]: and similarly for the other fields

File: Source/FieldSolver/WarpXExternalEMFields.cpp [hunk 4]
```diff
@@ -3,43 +3,53 @@
...
 
 using namespace amrex;
 
+/**
+ * \brief which_variable determines which field component the external excitation is applied on
+ * which_variable = 0 : external field excitation applied to all three field components, E, B and H
+ * which_variable = 1 : external field excitation applied to E field only
+ * which_variable = 2 : external field excitation applied to B field only
+ * which_variable = 3 : external field excitation applied to H field only; this option is only valid when USE_LLG == TRUE
+ */
```
Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXExternalEMFields.cpp:16]:
```

Comment (RevathiJambunathan) [inline Source/FieldSolver/WarpXExternalEMFields.cpp]:
```suggestion

```

File: Source/Evolve/WarpXEvolve.cpp [hunk 5]
```diff
@@ -380,9 +383,11 @@ WarpX::OneStep_nosub (Real cur_time)
...
         FillBoundaryG(guard_cells.ng_FieldSolverG);
 #ifndef WARPX_MAG_LLG
-        EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
-        if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
-        FillBoundaryB(guard_cells.ng_FieldSolver);
+            EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
+            if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
+            FillBoundaryB(guard_cells.ng_FieldSolver);
+            // ApplyExternalFieldExcitation
+            ApplyExternalFieldExcitationOnGrid("bfieldexternal"); // apply B external excitation; soft source to be fixed
```
Comment (ajnonaka) [inline Source/Evolve/WarpXEvolve.cpp]:
```suggestion
         EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
         if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
         FillBoundaryB(guard_cells.ng_FieldSolver);
         // ApplyExternalFieldExcitation
         ApplyExternalFieldExcitationOnGrid("bfieldexternal"); // apply B external excitation; soft source to be fixed
```

File: Source/Evolve/WarpXEvolve.cpp [hunk 6]
```diff
@@ -380,9 +383,11 @@ WarpX::OneStep_nosub (Real cur_time)
...
         FillBoundaryG(guard_cells.ng_FieldSolverG);
 #ifndef WARPX_MAG_LLG
-        EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
-        if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
-        FillBoundaryB(guard_cells.ng_FieldSolver);
+         EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
+         if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
+         FillBoundaryB(guard_cells.ng_FieldSolver);
+         // ApplyExternalFieldExcitation
+         ApplyExternalFieldExcitationOnGrid("bfieldexternal"); // apply B external excitation; soft source to be fixed
```
Comment (ajnonaka) [inline Source/Evolve/WarpXEvolve.cpp]:
```suggestion
        EvolveB(0.5_rt * dt[0]); // We now have B^{n+1/2}
        if (do_silver_mueller) ApplySilverMuellerBoundary( dt[0] );
        FillBoundaryB(guard_cells.ng_FieldSolver);
        // ApplyExternalFieldExcitation
        ApplyExternalFieldExcitationOnGrid("bfieldexternal"); // apply B external excitation; soft source to be fixed
```

File: Source/Utils/WarpXAlgorithmSelection.cpp [hunk 7]
```diff
@@ -93,6 +93,16 @@ const std::map<std::string, int> ParticleBCType_algo_to_int = {
     {"default",    ParticleBoundaryType::Absorbing}
 };
 
+const std::map<std::string, int> ExternalFieldType_algo_to_int = {
+    {"all", ExternalFieldType::All},
```
Comment (ajnonaka) [inline Source/Utils/WarpXAlgorithmSelection.cpp]: Can we change this to "allexternal" and "ExternalFIeldType::AllExternal"
using just 'all' is a little too generic and will be hard to search for

File: Source/Utils/WarpXAlgorithmSelection.H [hunk 8]
```diff
@@ -118,6 +118,19 @@ struct ParticleBoundaryType {
     };
 };
 
+/** Field component type to apply the external field excitation on
+ */
+struct ExternalFieldType {
+    enum {
+        All = 0,
```
Comment (ajnonaka) [inline Source/Utils/WarpXAlgorithmSelection.H]: AllExternal ?

File: Source/Evolve/WarpXEvolve.cpp (hunk 3 above)
Comment (ajnonaka) [reply Source/Evolve/WarpXEvolve.cpp]: And use "ExternalFieldType::AllExternal" (similar comments below)

**Review by ajnonaka — APPROVED** [review]

**Review by RevathiJambunathan — APPROVED** [review]

**Comment by RevathiJambunathan** [comment]
Thank you for the PR!

**Pull request merged**